sys/sched_round_robin/sched_round_robin.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2021 TUBA Freiberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       Round Robin Scheduler implementation
+ *
+ * @author      Karl Fessel <karl.fessel@ovgu.de>
+ *
+ * @}
+ */
+
+#include "sched.h"
+#include "thread.h"
+#include "xtimer.h"
+#include "sched_round_robin.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static void _sched_round_robin_cb(void *d);
+
+static xtimer_t _rr_timer = { .callback = _sched_round_robin_cb };
+
+/*
+ * assuming simple reads from and writes to a byte to be atomic on every board
+ */
+static uint8_t _current_rr_priority = 0;
```

Comment:
I'm not fully convinced that this variable is actually needed. Wouldn't it be enough to start and stop the timer based on whether the prio of the currently running thread is enabled for round robin and whether there is at least another thread waiting in the runq?

It certainly wouldn't hurt to verify in the timer callback that the currently running thread is indeed of eligible prio, just in case the timer triggered just before it would have been stopped.

Reply:
the problem this solves is not restarting (stop if prio changes(higher) and start when it changes back(lower)) the timer for a new interval if a short higher priority thread intervenes the current parralleled priority (eg two apps running, but a driver thread needs some cpu time)

read apps as: same priority just above idle threads

Reply:
Just to make sure we're on the same page:

- Let's assume we have high priority thread *t1* and low priority threads *t2* and *t3*
- *t2* and *t3* are scheduled round robin with a quantum of 1 ms
- Let's say *t1* wakes up every 750 us for a duration of 10 us
- So: *t2* gets preempted by *t* roughly 3/4 into its quantum
- When *t1* blocks again, *t2* gets restarted with a full quantum
- Hence, *t3* will starve as *t2* gets a new quantum after every change of runqueues, which happens consistently before *t3* get scheduled

Why not instead shuffle the new runqueue on a change of the active runqueue prior to scheduling the head of the queue? This way *t2* and *t3* would get a quantum of ~ 740 us in terms in the above scenario.

I agree that fairness is still an open issue with this approach. But it also is with the current PR.

Reply:
an than there is something with high priority that needs 5µ every 100µs